gossiper: load_saved_endpoint: set dc, rack, and shutdown status

bhalevy opened the pull request:
When loading endpoint_state from system.peers, pass the loaded nodes dc/rack info from storage_service::join_token_ring to gossiper::add_saved_endpoint.

Load the endpoint DC/RACK information to the endpoint_state, if available, and initialize the endpoint_state STATUS to `shutdown` if it owns tokens, as this is the best we know about the endpoint at this stage.

Fixes scylladb/scylladb#15787

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4302/
- Duration: 2 hr 26 min
- Builder: spider4.cloudius-systems.com

bhalevy commented on gms/gossiper.cc:
```diff
@@ -2120,15 +2120,20 @@ future<> gossiper::add_saved_endpoint(inet_address ep) {
     if (!tokens.empty()) {
         std::unordered_set<dht::token> tokens_set(tokens.begin(), tokens.end());
         ep_state.add_application_state(gms::application_state::TOKENS, versioned_value::tokens(tokens_set));
+        ep_state.add_application_state(gms::application_state::STATUS, gms::versioned_value::shutdown(true));
```
If it is removed, how come it owns tokens in system.peers?

kbr-scylla replied:
It could happen like this:
1. shutdown nodes A, B
2. on node C perform `--removenode A --ignore-dead-nodes B`
3. node A is removed, but on node B it still owns tokens.
4. Restart node B. Node A gets resurrected.

The question is, if some node was used inside `--ignore-dead-nodes`, maybe we should assume it is also permanently dead.

tgrabiec replied:
> The question is, if some node was used inside --ignore-dead-nodes, maybe we should assume it is also permanently dead.

dead = permanently down

It's only safe to ignore a node if it's permanently down.

bhalevy replied:
We can also extend removenode to remove multiple nodes at once, but I'm not sure we want to go there.

bhalevy replied:
Another alternative is to provide an api to declare nodes as DEAD before removenode or replace.

bhalevy replied:
This can be record persistently and be used to fence them off permanently.

bhalevy commented:
@asias please ack

bhalevy commented:
ping @asias

asias commented:
> @asias please ack

Per the discussion in the past (https://github.com/scylladb/scylladb/issues/15787#issuecomment-1774320071), I do not think each of the nodes should inject the status=shutdown for other nodes. 

I proposed here: 

https://github.com/scylladb/scylladb/issues/15787#issuecomment-1776789517

to let the replacing node to inject the status=normal/shutdown when it uses the --ignore-dead-nodes-for-replace.

mykaul commented:
@kbr-scylla - please review so we can get this in (and backported)

bhalevy commented:
Note that this fix works only with gossip based replace.
With consistent-cluster-management the replacing node, when ignoring two other dead nodes, after cluster restarts.
Goes into a loop like this one:
```
INFO  2023-12-06 09:33:16,077 [shard 0:strm] raft_group0 - server e17fc0ce-3ed1-4ef1-9f0d-df4089af47ef found group 0 with group id 6ea1ad50-9409-11ee-9cfe-caa034bf2eaa, leader cc6648fa-2417-4bdb-8657-d8454e82a586
WARN  2023-12-06 09:33:17,077 [shard 0:strm] raft_group0 - failed to modify config at peer cc6648fa-2417-4bdb-8657-d8454e82a586: seastar::rpc::timeout_error (rpc call timed out). Retrying.
```

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5086/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6447/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5086/)
#### Failed Tests (6/22003):
- [test_replace_node_no_hibernate_state[use_host_id-rbo_disabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5086/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest___test_replace_node_no_hibernate_state_use_host_id_rbo_disabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_node_no_hibernate_state[use_host_id-rbo_disabled])
- [test_replace_node_no_hibernate_state[use_endpoint-rbo_disabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5086/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest___test_replace_node_no_hibernate_state_use_endpoint_rbo_disabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_node_no_hibernate_state[use_endpoint-rbo_disabled])
- [test_change_node_ip_full_cluster_down](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5086/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_change_node_ip_full_cluster_down) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_node_ip_full_cluster_down)
- [test_change_node_ip_full_cluster_down](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5086/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_change_node_ip_full_cluster_down_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_node_ip_full_cluster_down)
- [test_replace_node_no_hibernate_state[use_host_id-rbo_enabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5086/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest___test_replace_node_no_hibernate_state_use_host_id_rbo_enabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_node_no_hibernate_state[use_host_id-rbo_enabled])
- [test_replace_node_no_hibernate_state[use_endpoint-rbo_enabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5086/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest___test_replace_node_no_hibernate_state_use_endpoint_rbo_enabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_node_no_hibernate_state[use_endpoint-rbo_enabled])
#### Build Details:
- Duration: 2 hr 11 min
- Builder: spider5.cloudius-systems.com

kbr-scylla commented:
`consistent_cluster_management` is still using "gossip-based replace", it only adds an additional step to the procedure, which is modifying group 0 configuration.

Raft-based topology operations are only enabled under the experimental feature `consistent-topology-changes`.

kbr-scylla commented:
>  With consistent-cluster-management the replacing node, when ignoring two other dead nodes, after cluster restarts.
> Goes into a loop like this one:

How many nodes are there in the cluster? Perhaps you're trying to do a topology change without a majority, which is impossible when Raft is enabled. 

https://opensource.docs.scylladb.com/master/architecture/raft.html#quorum-requirement

bhalevy commented:
> > With consistent-cluster-management the replacing node, when ignoring two other dead nodes, after cluster restarts.
> > Goes into a loop like this one:
> 
> How many nodes are there in the cluster? Perhaps you're trying to do a topology change without a majority, which is impossible when Raft is enabled.
> 
> https://opensource.docs.scylladb.com/master/architecture/raft.html#quorum-requirement
> 
> EDIT: What is the exact scenario? Are you talking in the context of some particular test?

Yes, the except above relates to repair_based_node_operations_test.py::TestRepairBasedNodeOperations::test_ignore_dead_nodes_for_replace_option_multi_rack (from https://github.com/scylladb/scylla-dtest/pull/3685/commits/7191708f28d2c2203ba762f5b0929cb88dd1f8c6)
It starts a 6 node cluster, single DC, 2 nodes per rack.
Then it stops the whole cluster and restarts only 3 nodes out of the 6, and this apparently prevents the quorum to form.
It would be clearer if group0 would have printed a clearer message to the log that it's waiting for a quorum of the nodes.

bhalevy commented:
We can configure the test cluster with 3 nodes in the first rack so to have 4 nodes out of 7 post restart.

kbr-scylla commented:
>  It would be clearer if group0 would have printed a clearer message to the log that it's waiting for a quorum of the nodes.

Agreed, but nobody sent a formal complaint yet (i.e., created an issue). I often forget about it in heat of the moment when people casually mention it in an unrelated thread

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -2388,7 +2388,7 @@ void gossiper::force_newer_generation() {
     eps.get_heart_beat_state().force_newer_generation_unsafe();
 }
```
`storage_service: load_saved_endpoint of token owners with unknown state`

> However, a bootstrapping node, has no persistent knowledge of the down nodes,
> and it learns about their existance from gossip.
> But, since the down nodes have unknown status, they never go
> through `handle_state_normal`.

You mean any starting node (including restarts), not just bootstrapping?

kbr-scylla replied:
Wasn't the original idea to inject STATUS of peers? I'm confused. The PR or commit messages don't seem to explain the motivation behind this approach.

bhalevy replied:
> You mean any starting node (including restarts), not just bootstrapping?

No, restarting node that already joined the cluster will load the down-nodes state from their local system.peers table.

bhalevy replied:
> Wasn't the original idea to inject STATUS of peers? I'm confused. The PR or commit messages don't seem to explain the motivation behind this new approach.

That was the original idea, but per https://github.com/scylladb/scylladb/issues/15787#issuecomment-1774320071
this is not deemed safe according to @asias.

kbr-scylla commented:
> The rest of this series deals with loading the saved states also from storage_service::on_join for all token owners
and also in the replace path, from storage_service::prepare_replacement_info.

By "all token owners" you also mean non-NORMAL ones (e.g. bootstrapping nodes)?
Why?

bhalevy commented on service/storage_service.cc:
```diff
@@ -4087,47 +4095,43 @@ future<> storage_service::join_cluster(sharded<db::system_distributed_keyspace>&
+            } else if (!st.tokens.empty()) {
+                tmptr->update_topology(ep, get_dc_rack(st), locator::node::state::normal);
+                co_await tmptr->update_normal_tokens(st.tokens, ep);
+                if (st.host_id) {
```
Agreed. system.peers or system.topology should always have a valid host_id.
I'll also add a warning about it if it's missing in 
system_keyspace::load_host_ids.
Just note that `update_peer_info` adds the columns one at a time, so if scylla crashes while adding a node we might end up with an incomplete row for it.
To fix that we may want to add the host_id always last and ignore the row if it doesn't contain a host_id.

bhalevy replied:
I'll submit another PR for enforcing host_id in system.peers

bhalevy replied:
https://github.com/scylladb/scylladb/pull/16376

bhalevy commented on service/storage_service.cc:
```diff
@@ -3900,9 +3900,9 @@ future<> storage_service::on_change(inet_address endpoint, application_state sta
             co_return;
         }
+        slogger.debug("endpoint={} on_change:     updating system.peers table", endpoint);
+        co_await do_update_system_peers_table(endpoint, state, value);
```
Yeah, I need to expand on the motivation.
Since @asias insists on not initializing the peer status (to "shutdown")
the peers that are down after full cluster restart never become normal token owners,
hence this change.

> system.peers is supposed to contain only normal nodes.

Is it?
I think it contains nodes while they are leaving too.

It's be easier if we kept the peer state there too
and manage the table in a distributed way over raft :)

> Aren't you turning bootstrapping nodes into normal nodes prematurely this way?

I need to swap this patch with the next one:
`storage_service: on_join: defer on_change for STATUS and other states`
so that the status will be handled last and saved only if the endpoint is a normal token owner is we do today.

I think it should ok to have endpoints in system.peers with no STATUS while they bootstrap, but I don't know if this is cleaned up if bootstrapping failed midway.
@asias, @tgrabiec what do you think?

bhalevy replied:
Turns out we can do without this patch.
I'll drop it in the next version.

bhalevy replied:
But you got a good point about adding the state prematurally.
It interferes with replace, as seen in the `test_replace_node_no_hibernate_state` dtest.

bhalevy replied:
I dropped this logic from the next version.
Instead, the endpoint state of the explicitly ignored nodes will be used in prepare_replacement_info to populate topology and token_metadata

bhalevy commented on service/storage_service.cc:
```diff
@@ -3849,8 +3849,31 @@ future<> storage_service::handle_state_removed(inet_address endpoint, std::vecto
 
 future<> storage_service::on_join(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, gms::permit_id pid) {
```
I'm not sure how the existing code was supposed to work but handling of `RPC_READY` and `INTERNAL_IP` is done in conditional block, under `is_normal_token_owner(endpoint)`.
This change makes sure `STATUS` is handled before them so that an endpoint can become a normal token owner after handling `STATUS` by `on_change`, and after that `on_change` is called for `RPC_READY` and `INTERNAL_IP`.

I tried to explain that in the commit message, but apparently, unsuccessfully.
I'll try to clarify it.

bhalevy replied:
dropped from next version

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5175/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6564/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5175/)
#### Build Details:
- Duration: 2 hr 19 min
- Builder: spider7.cloudius-systems.com

bhalevy commented on service/storage_service.cc:
```diff
@@ -4087,47 +4095,43 @@ future<> storage_service::join_cluster(sharded<db::system_distributed_keyspace>&
-        for (auto x : loaded_tokens) {
-            auto ep = x.first;
-            auto tokens = x.second;
+            for (const auto& [ep, st] : loaded_endpoints) {
```
Fixed in next version

bhalevy commented:
In v3 ([1bccb5f](https://github.com/scylladb/scylladb/commit/1bccb5fb669a5bf6b0a278d613bb3d03a805cb1d))
- Simplified the series overall
- locator: host_id_or_endpoint: support hash and fmt::format
  - for keeping a `std::unordered_set<locator::host_id_or_endpoint>` in `replacement_info.ignore_nodes`
- gms: endpoint_state: add getters for host_id, dc_rack, and tokens
  - to simplify next patch
- storage_service: prepare_replacement_info: load ignored nodes state

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -2087,20 +2087,20 @@ future<> gossiper::add_saved_endpoint(inet_address ep) {
+        ep_state.add_application_state(gms::application_state::RACK, versioned_value::datacenter(loaded_state.opt_dc_rack->rack));
     }
     auto generation = ep_state.get_heart_beat_state().get_generation();
+    logger.debug("Adding saved endpoint {} generation={} version={}: {}", ep, generation, ep_state.get_heart_beat_state().get_heart_beat_version(), ep_state);
```
Generation will be always 0 here, right?

bhalevy replied:
If the endpoint state already exists we reuse it so generation might not be zero.
That was the existing logic.  I'm not sure if it's really required.

tgrabiec replied:
If the endpoint exists, we call:
```
ep_state.set_heart_beat_state_and_update_timestamp(heart_beat_state());
```

It would be wrong to set a state in the current generation on behalf of another node using our local version generator, so it must be set with generation 0 for correctness.

bhalevy replied:
You're right. It should reuse `ep_state.get_heart_beat_state()` instead of the local `heart_beat_state`, as done in `apply_new_states`, so it just updates the timestamp and keeps the heartbeat_state.

I opened https://github.com/scylladb/scylladb/issues/16429 and will include the fix in the patchset so it can be backported independently if we want to.

bhalevy replied:
https://github.com/scylladb/scylladb/pull/16432

bhalevy commented on service/storage_service.cc:
```diff
@@ -4479,15 +4480,48 @@ storage_service::prepare_replacement_info(std::unordered_set<gms::inet_address>
+            if (res.size() == 0) {
+                throw std::runtime_error(::format("Could not find ignored node with host_id {}", hioa.id));
+            } else if (res.size() > 1) {
+                throw std::runtime_error(::format("Find multiple nodes to ignore with host_id {}: {}", hioa.id, res));
```
fixed in v4

bhalevy commented:
In v4 ([c4ab97f](https://github.com/scylladb/scylladb/commit/c4ab97f623a348c1c2b9da40a643d1f3353e8813)):
- rebased on top of 26cbd28883edb31e2d4db92bcbf56e0019ac7b23
- includes a fix for https://github.com/scylladb/scylladb/issues/16429 (and submitted separately)
- fixed grammar in runtime_error message.

Draft until https://github.com/scylladb/scylladb/pull/16432 and until we resolve https://github.com/scylladb/scylladb/pull/15788#discussion_r1421742652

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5278/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6686/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5278/)
#### Build Details:
- Duration: 4 hr 6 min
- Builder: spider5.cloudius-systems.com

asias commented on service/storage_service.cc:
```diff
@@ -559,7 +560,13 @@ future<> storage_service::topology_state_load() {
-            co_await _gossiper.add_saved_endpoint(ep);
+            gms::loaded_endpoint_state st;
+            const auto& topo = tmptr->get_topology();
+            const auto* node = topo.find_node(e);
```
Can node be nullptr here?

bhalevy replied:
It'd be a bug.
If it's in token_metadata then it must be in topology.  They are tightly coupled.  That's why topology is a member of t.m. and is manipulated through it, rather than being a standalone class, like raft topology.

bhalevy replied:
I added a comment explaining this in next version

asias commented on gms/gossiper.cc:
```diff
@@ -2088,22 +2092,20 @@ future<> gossiper::add_saved_endpoint(inet_address ep) {
         ep_state = *es;
         logger.debug("not replacing a previous ep_state for {}, but reusing it: {}", ep, ep_state);
-        ep_state.set_heart_beat_state_and_update_timestamp(heart_beat_state());
+        ep_state.update_timestamp();
```
We should avoid using non-zero generation and inject gossip sates. The code before this change guarantees it. We can check if the current ep_state already has the dc / rack / tokens /host id, skip injecting if so.  Or, we can reuse the existing  ep_state but reset the generation to zero. It is dangerous to reuse non-zero generation gossip state + injecting. 

Are we calling add_saved_endpoint when we have collected all the states loaded_state and only call it once during startup? We should, this will simplify how we inject.

bhalevy replied:
> We should avoid using non-zero generation and inject gossip sates. The code before this change guarantees it.

How?

Please see latest version of https://github.com/scylladb/scylladb/pull/16432. It won't update an existing ep_state if it has non-zero generation.

asias replied:
Because `ep_state.set_heart_beat_state_and_update_timestamp(heart_beat_state())` always set the generation to zero.

bhalevy replied:
It's better to skip loading local state if there is an existing endpoint state with non-zero generation.

bhalevy replied:
The reason is that resetting the ep_state generation to 0 will cause any update from remote with non-zero generation to win over the local endpoint state and replace it anyhow, so a non-zero generation means we should lose the race here too and not bother with updating the endpoint state..

bhalevy commented:
In v5 ([9c0f042](https://github.com/scylladb/scylladb/commit/9c0f042a542cb56e102ef02b6f68db8bf7c2820a))
- rebased
- clarified why node must be found in topology when is `get_all_endpoints`

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5593/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6993/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5593/)
#### Failed Tests (4/23684):
- [test_raft_replace_ignore_nodes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5593/testReport/junit/%28root%29/test_raft_ignore_nodes/Tests___Unit_Tests___test_raft_replace_ignore_nodes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_replace_ignore_nodes)
- [topology_experimental_raft.test_raft_ignore_nodes.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5593/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_raft_ignore_nodes_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_raft_ignore_nodes.dev.3)
- [test_raft_remove_ignore_nodes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5593/testReport/junit/%28root%29/test_raft_ignore_nodes/Tests___Unit_Tests___test_raft_remove_ignore_nodes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_remove_ignore_nodes)
- [topology_experimental_raft.test_raft_ignore_nodes.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5593/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_raft_ignore_nodes_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_raft_ignore_nodes.dev.1)
#### Build Details:
- Duration: 1 hr 54 min
- Builder: spider8.cloudius-systems.com

bhalevy commented:
rebased

bhalevy commented:
@tgrabiec @asias please re-review

dani-tweig commented:
@bhalevy , any update?

bhalevy commented:
> @bhalevy , any update?

We're awaiting review from @tgrabiec and @asias

dani-tweig commented:
@bhalevy Can you share an ETA?

tgrabiec commented on locator/token_metadata.hh:
```diff
@@ -390,4 +394,25 @@ public:
+
+    template <typename FormatContext>
+    auto format(const locator::host_id_or_endpoint& hioa, FormatContext& ctx) const {
+        return fmt::format_to(ctx.out(), "{}/{}", hioa.id, hioa.endpoint);
```
If it's host_id OR endpoint, then we should print either id or endpoint.

bhalevy replied:
It's an inclusive OR.  Both can be present.
Otherwise, a std::variant could be used as well.  Printing both would allow us to see the full state.
We can becsmarter and print only one if them if the other is missing, or both in case both are present.

bhalevy commented:
In v6 ([e4ebff0](https://github.com/scylladb/scylladb/commit/e4ebff0135bb577376d12c0f913decfdfb532333)):
- rebased
- fixed host_id_or_endpoint formatter as per https://github.com/scylladb/scylladb/pull/15788#discussion_r1462512037
- Added:
  - gossiper: start_gossiping: log local state
  - storage_service: topology_state_load: set local STATUS state using add_saved_endpoint

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6014/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7341/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6014/)
#### Failed Tests (72/23831):
- [test_fetch_from_system_tables](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6014/testReport/junit/%28root%29/test_system_tables/Tests___Unit_Tests___test_fetch_from_system_tables) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_fetch_from_system_tables)
- [test_fetch_from_system_tables](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6014/testReport/junit/%28root%29/test_system_tables/Tests___Unit_Tests___test_fetch_from_system_tables_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_fetch_from_system_tables)
- [test_fetch_from_system_tables](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6014/testReport/junit/%28root%29/test_system_tables/Tests___Unit_Tests___test_fetch_from_system_tables_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_fetch_from_system_tables)
- [test_repair_task_progress_finished_task](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6014/testReport/junit/%28root%29/test_repair_task/Tests___Unit_Tests___test_repair_task_progress_finished_task) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_task_progress_finished_task)
- [test_repair_task_tree](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6014/testReport/junit/%28root%29/test_repair_task/Tests___Unit_Tests___test_repair_task_tree) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_task_tree)
- [test_repair_task_progress](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6014/testReport/junit/%28root%29/test_repair_task/Tests___Unit_Tests___test_repair_task_progress) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_task_progress)
- [test_repair_task_progress_finished_task](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6014/testReport/junit/%28root%29/test_repair_task/Tests___Unit_Tests___test_repair_task_progress_finished_task_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_task_progress_finished_task)
- [test_repair_task_tree](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6014/testReport/junit/%28root%29/test_repair_task/Tests___Unit_Tests___test_repair_task_tree_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_task_tree)
- [test_repair_task_progress](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6014/testReport/junit/%28root%29/test_repair_task/Tests___Unit_Tests___test_repair_task_progress_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_task_progress)
- [test_repair_task_progress_finished_task](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6014/testReport/junit/%28root%29/test_repair_task/Tests___Unit_Tests___test_repair_task_progress_finished_task_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_task_progress_finished_task)
#### Build Details:
- Duration: 2 hr 5 min
- Builder: spider8.cloudius-systems.com

tgrabiec commented on locator/token_metadata.hh:
```diff
@@ -58,6 +58,10 @@ struct host_id_or_endpoint {
     host_id_or_endpoint(const sstring& s, param_type restrict = param_type::auto_detect);
 
+    bool operator==(const host_id_or_endpoint& o) const noexcept {
+        return has_host_id() ? id == o.id : endpoint == o.endpoint;
```
If both host id and endpoint can be set then this operator is not symmetric, which can lead to undefined behavior.

Example:

a = {null, 10.0.0.1}
b = {id1, 10.0.0.1}

a == b is true
b == a is false

tgrabiec replied:
We should document that "or" in the name is not an exclusive "or", it's not obvious.

bhalevy replied:
maybe and_or would be better then, though it's more cumbersome

bhalevy replied:
> If both host id and endpoint can be set then this operator is not symmetric, which can lead to undefined behavior.

Will fix

bhalevy replied:
fixed and renamed in v7

bhalevy commented:
In v7 ([8b3fa53](https://github.com/scylladb/scylladb/commit/8b3fa538830a80c42443f51fe33b974eac410510)):
- storage_service: topology_state_load: set local STATUS state using add_saved_endpoint
  - restored hunk that was deleted by mistake, causing failure in topology_custom/test_maintenance_mode
- locator: host_id_or_endpoint: support hash and fmt::format
  - fixed asymmetrical equality comparison
  - added unit test for equality op
- locator: rename host_id_or_endpoint to *and_or*

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6283/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6283/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6283/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7581/)
#### Build Details:
- Duration: 2 hr 1 min
- Builder: spider7.cloudius-systems.com

bhalevy commented on locator/token_metadata.hh:
```diff
@@ -388,4 +392,31 @@ public:
+template<>
+struct hash<locator::host_id_or_endpoint> {
+    size_t operator()(const locator::host_id_or_endpoint& i) const {
+        return i.has_host_id() ? hash<locator::host_id>()(i.id) : hash<gms::inet_address>()(i.endpoint);
```
right again.  duh, I should've thought about that.
thanks

bhalevy replied:
Equality/hashing by host_id, regardless of address is not a strict requirement.
So I'm dropping that, and implementing simple and symmetrical equality and hash function
that take both host_id and endpoint into account.

bhalevy commented:
In v8 ([c8bd8f3](https://github.com/scylladb/scylladb/commit/c8bd8f395f7bc33ae19649181fde47174652d4de)):
- rebased
- simplified host_id_and_or_endpoint equality op and std::hash function
  - and made them symmetrical, taking both host_id and endpoint into account.

avikivity commented on gms/gossiper.cc:
```diff
@@ -2026,11 +2026,9 @@ future<> gossiper::start_gossiping(gms::generation_type generation_nbr, applicat
 
-    co_await replicate(get_broadcast_address(), std::move(local_state), permit.id());
-
-    logger.trace("gossip started with generation {}", generation);
```
Shouldn't we keep the generation (at least in trace)? @asias

bhalevy replied:
It's printed below as part of `local_state`.
Previously `local_state` was moved to `replicate` so we needed to copy `generation` first before moving it, but this change keeps it so we can print the whole endpoint_state, not just the generation.

avikivity replied:
ok

avikivity commented on gms/gossiper.cc:
```diff
@@ -2134,12 +2134,16 @@ void gossiper::build_seeds_list() {
+    if (!loaded_state.host_id) {
+        on_internal_error(logger, format("can't find host_id for ep {} for adding saved endpoint", ep));
+    }
+
```
Could talking to an older node cause this to trigger?

bhalevy replied:
Not because it's old, since we had set the host_id in system.peers forever.
Only an exception in this process could have left the state in system.peers with no host_id (see https://github.com/scylladb/scylladb/pull/16376).
If we somehow hit this internal error, we should fix system.peers manually.

avikivity replied:
ok

avikivity commented on gms/gossiper.cc:
```diff
@@ -2154,24 +2158,18 @@ future<> gossiper::add_saved_endpoint(inet_address ep, permit_id pid) {
+    ep_state.add_application_state(gms::application_state::HOST_ID, versioned_value::host_id(loaded_state.host_id));
+    if (!loaded_state.tokens.empty()) {
+        ep_state.add_application_state(gms::application_state::TOKENS, versioned_value::tokens(loaded_state.tokens));
+    }
```
Is making these parts unconditional related to the rack/dc? If so please explain in the changelog.

bhalevy replied:
okay, I'll add an explanation.
Looking up the host_id is up to the caller now.
And basically the tokens, dc, and rack are useless without host_id.

bhalevy replied:
The host_id can be made mandatory after 50ebce8acc2406bcc8466b2f051e2ceae395e83e

avikivity commented on gms/endpoint_state.cc:
```diff
@@ -61,6 +62,33 @@ future<> i_endpoint_state_change_subscriber::on_application_state_change(inet_ad
+    if (auto* app_state = get_application_state_ptr(application_state::TOKENS)) {
+        ret = versioned_value::tokens_from_string(app_state->value());
+    }
+    return ret;
```
Should we return optional here too? so before using it, you must check if it's engaged.

avikivity commented on service/storage_service.cc:
```diff
@@ -2756,15 +2757,51 @@ storage_service::prepare_replacement_info(std::unordered_set<gms::inet_address>
         .address = replace_address,
     };
+
+    auto ignore_hioas = parse_node_list(_db.local().get_config().ignore_dead_nodes_for_replace());
```
what is a hioas?

bhalevy replied:
It was supposed to be related to host_id and/or endpoint (or address), but it's a bad name
and I'm changing it in the next version.
Essentially I'm simplifying host_id_or_endpoint to keep a std::variant value that can be either host_id or endpoint, but not both.

avikivity commented on service/storage_service.cc:
```diff
@@ -2756,15 +2757,51 @@ storage_service::prepare_replacement_info(std::unordered_set<gms::inet_address>
+        tmptr->update_topology(hioa.id, opt_dc_rack);
+        if (hioa.id) {
+            tmptr->update_host_id(hioa.id, hioa.endpoint);
+        }
```
Will all these throws trigger in production and kill clusters?

bhalevy replied:
They indicate an operator error related to the optional ignore-nodes values.
They won't kill the cluster, but rather kill the replacing node, preventing it from joining, which is what they're supposed to do as the ignore nodes host_id:s must refer to the dead nodes that are no longer in the cluster.
Unfortunately, we still must know their assigned ip address until gossiper transitions to use host_id:s.
In the future we can support host_id only and drop the support for identifying ignore-nodes using ip addresses.

Cc @tgrabiec

avikivity commented:
How does this interact with consistent topology?

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6360/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6360/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7638/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6360/)
#### Build Details:
- Duration: 2 hr 3 min
- Builder: spider8.cloudius-systems.com

bhalevy commented:
> How does this interact with consistent topology?

- Initial state loaded by `storage_service::topology_state_load`.
- For the replacing node, `prepare_replacement_info` now loads both the replaced and and the ignored nodes state in a common path, both for consistent and legacy topology changes

bhalevy commented:
In v9 ([293b1b6](https://github.com/scylladb/scylladb/commit/293b1b6092bdc6c481b0f77397ce02aa56c509b7)):
- rebased
- gossiper: load_saved_endpoint: set dc and rack
  - expanded in commit log about the making host_id mandatory
- gms: endpoint_state: add getters for host_id, dc_rack, and tokens
  - made get_tokens return std::optional so that caller can know if the app state is missing or if it contains empty tokens list (makes no real functional change in this series, but could be useful in the future)
- locator: host_id_or_endpoint: keep value as variant
  - dropped the std::hash and fmt stuff for `host_id_or_endpoint` and made it simpler altogether
  - simplified naming of related variables
- storage_service: prepare_replacement_info: load ignored nodes state
  - used `std::list` for ignored_nodes rather than `unordered_set` since `host_id_or_endpoint` has deleted default ctor
  - used `std::unordered_map<locator::host_id, gms::inet_address>` for `replacement_info.ignore_nodes` for the above reason and for future compatibility once we always use the host_id:s (today we still need to collect the inet_address values for node ops)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6430/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6430/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7677/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6430/)
#### Build Details:
- Duration: 1 hr 52 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
@tgrabiec / @avikivity please re-review

bhalevy commented:
rebased.

> @tgrabiec / @avikivity please re-review

ping review

avikivity commented:
> The first part of this series loads also DC and RACK from system.peers to make them available to the replacing node as they are crucial for building a correct replication map with network topology replication strategy.

How did the system work without this?

avikivity commented on gms/endpoint_state.cc:
```diff
@@ -61,6 +62,32 @@ future<> i_endpoint_state_change_subscriber::on_application_state_change(inet_ad
+    if (auto* app_state = get_application_state_ptr(application_state::HOST_ID)) {
+        return locator::host_id(utils::UUID(app_state->value()));
+    }
+    return locator::host_id::create_null_id();
```
Is null id the best choice here? It's too easy for it to slip through and get used as if it is a real ID. Shouldn't we preserve it as optional so it's clear that it's unser?

bhalevy replied:
I don't think that's required, a null host_id is never valid and it generally stands for an absent host_id.
returning `std::optional<locator::host_id>` is an over-kill IMO.
The system would behave practically the same is std::nullopt is returned, or a std::optional<host_id> holding a null id is returned.  Both cases are invalid and shouldn't happen.

kbr-scylla replied:
A slight inconsistency with `get_tokens()`, which could return an empty set instead of nullopt following the same reasoning.

bhalevy replied:
In a previous version it did just that, but Avi asked to return nullopt to be able to distinguish the two cases.
In theory, we may want to keep a node with TOKENS application state containing an empty list (maybe when we deprecate vndoes entirely and rely only on tablets), but a node cannot have a null host_id.

bhalevy replied:
correction: we don't allow empty tokens list (see `system_keyspace::load_topology_state`)

bhalevy replied:
@avikivity, based on https://github.com/scylladb/scylladb/blob/57c408ab5d206980fece198a1278e5612629bf08/db/system_keyspace.cc#L2706-L2710, we cannot have empty tokens in system.peers, so we don't really need to return an optional set of tokens in endpoint_state::get_tokens, we can safely return an empty set instead if the application state is missing.

avikivity commented on locator/token_metadata.hh:
```diff
@@ -45,8 +45,7 @@ class token_metadata;
 struct host_id_or_endpoint {
-    host_id id;
-    gms::inet_address endpoint;
+    std::variant<host_id, gms::inet_address> _value;
```
This looks fishy. Do we have a case where we know the host id, and a separate case where we only know the IP address?

Can't we only store the host id?

avikivity replied:
Another way to look at it: can we find the host_id during construction, and store only that?

bhalevy replied:
it represents the values we get from the config option that can be either ip addresses or host_id:s, both we parse from strings that aren't guaranteed to represent valid nodes when they are parsed.
Here we "resolve" them by looking for nodes identified by these addresses.  This just holds the parsed options.

kbr-scylla replied:
BTW we need to deprecate IP address passing.

bhalevy replied:
> BTW we need to deprecate IP address passing.

Yes.  (out of scope for this PR)

avikivity commented on service/storage_service.cc:
```diff
@@ -1339,7 +1339,7 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
         if (_sys_ks.local().bootstrap_complete()) {
             throw std::runtime_error("Cannot replace address with a node that is already bootstrapped");
         }
```
> However, a bootstrapping/replacing node has no persistent knowledge
of the down nodes, and it learns about their existance only from gossip.
But since the down nodes have unknown status, they never go
through `handle_state_normal` and therefore they are not accounted
as normal token owners.

Is this true for raft topology too?

I think for Raft topology the node should first become a member of Raft, and so can learn everything about the cluster, then start the replacement process.

bhalevy replied:
This was true also for raft topology at the time.
With all the recent changes there I'm not 100% sure that it's still the case.
Like I proposed elsewhere to @kbr-scylla, I think that we should move the endpoint_state map from gossiper to raft topology. With consistent-topology-changes enabled, raft topology, and only it, should manage the topology-related state while the gossiper can update non-topology related states.

kbr-scylla replied:
but endpoint_state map is a gossiper specific structure. It has stuff like generation or heartbeat or versioned values, none of which are relevant for raft-topology.

kbr-scylla replied:
> Is this true for raft topology too?

It's not. In Raft-topology, joining node (that includes node who wants to replace another node) is performing a handshake with the cluster by sending RPC to one of existing nodes, and it's the handler of the handshake who deals with ignore_nodes, verifies correctness of the request, liveness of dead nodes and so on.

bhalevy replied:
> but endpoint_state map is a gossiper specific structure. It has stuff like generation or heartbeat or versioned values, none of which are relevant for raft-topology.

Once raft topology is assuming ownership of the topology-related states (tokens, dc, rack) we can't say it's gossip-specific anymore.  The most major question is who is determining the node liveliness and its existence in the endpoint_state map.  With consistent topology, I'm certain that a. the responsibility should in one place, not dispersed between the gossiper and raft-topology, and b. it should move to raft-topology based on group0 membership.  Gossip can still be used in this model for states that are not critical for raft-topology, like heat map, compression dictionaries, etc.

kbr-scylla replied:
> Once raft topology is assuming ownership of the topology-related states (tokens, dc, rack) we can't say it's gossip-specific anymore.

Then you're not talking about endpoint_state map, but certain states that are distributed through gossip, and gossip internally uses endpoint_state map but that's an implementation detail of gossiper.

>  The most major question is who is determining the node liveliness and its existence in the endpoint_state map.

Node **liveness** is and will be determined by a failure detector, not by consensus. The consensus algorithm itself depends on failure detection to progress (leader election)

avikivity commented:
I don't see any tests.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6762/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6762/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7985/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6762/)
#### Build Details:
- Duration: 1 hr 59 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
> I don't see any tests.

I added dtests that reproduce https://github.com/scylladb/scylladb/issues/15787 after a full cluster restart with down nodes.
Other than that, every replace test exercises this path.
So I don't think additional tests are requried.

bhalevy commented:
> > The first part of this series loads also DC and RACK from system.peers to make them available to the replacing node as they are crucial for building a correct replication map with network topology replication strategy.
> 
> How did the system work without this?

As long as there is a living node that holds the down node's endpoint state in-core, the down node state, including DC, RACK, HOST_ID, and STATUS are gossiped from the live node(s).
The system never work properly in a multi-dc cluster after full cluster restart when more than one node is down.
Well, before we implemented ignore-nodes-for-replace, we could replace only one node and if the replacing node was in the same DC and RACK as the replaced node everything would be fine.
It is the combination after full-cluster restart, more than one node is DOWN, and the replacing one of the dead nodes while ignoring the rest.

bhalevy commented:
@tgrabiec @kbr-scylla you are listed in "changes requested" above.
Please re-review and ack/nack.
I'd like to merge this PR and then work on the backports since we need it in the field.

bhalevy commented:
@avikivity is there anything that still needs fixing in this series?

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -3043,26 +3042,68 @@ storage_service::prepare_replacement_info(std::unordered_set<gms::inet_address>
+        }
+        auto opt_dc_rack = esp->get_dc_rack();
+        tmptr->update_topology(id, opt_dc_rack);
+        tmptr->update_host_id(id, endpoint);
```
`prepare_replacement_info` should not modify token_metadata. Previously it would return the obtained information to the caller, `join_token_ring`, who was responsible for updating token_metadata etc. Why not continue this pattern? And I still see that this is the case:
```cpp
        if (!raft_topology_change_enabled()) {
            bootstrap_tokens = std::move(ri->tokens);
            slogger.info("Replacing a node with {} IP address, my address={}, node being replaced={}",
                get_broadcast_address() == *replace_address ? "the same" : "a different",
                get_broadcast_address(), *replace_address);
            tmptr->update_topology(tmptr->get_my_id(), std::nullopt, locator::node::state::replacing);
            tmptr->update_topology(ri->host_id, std::move(ri->dc_rack), locator::node::state::being_replaced);
            co_await tmptr->update_normal_tokens(bootstrap_tokens, ri->host_id);
            tmptr->update_host_id(ri->host_id, *replace_address);
            replaced_host_id = ri->host_id;
        }
```

So first we update token_metadata here, and then again at the call site?

The more different places of updating state we have, the harder it is to understand the system.

Also, these updates should probably be disabled in raft-topology mode (like the updates I showed above), where responsibility of updating of token_metadata is shifted to group0 state machine.

bhalevy replied:
It's updating the state for the ignored nodes, not the replaced node.
Without that, with replace_ignore after full cluster restart the ignored nodes never go through handle_state_normal (since they are not considered normal token owners with a `NORMAL`  status in endpoint_state).

I'm ok with letting `join_token_ring` do that.
And I'll also test with consistent-topology-changes to make sure it's happening there too.

kbr-scylla replied:
> I'm ok with letting join_token_ring do that.

Yes please, let's put all the updates to token_metadata together in the code.

>  And I'll also test with consistent-topology-changes to make sure it's happening there too.

It should already be happening without this code, through `topology_state_load`, after joining node becomes a member of group 0. Once it performs the initial read_barrier it should include all nodes including the ignored ones in its token ring.

So we should be able to put this code in the `!raft_topology_change_enabled()` branch

bhalevy replied:
done in v10

bhalevy commented on gms/gossiper.cc:
```diff
@@ -2154,24 +2158,18 @@ future<> gossiper::add_saved_endpoint(inet_address ep, permit_id pid) {
-    // It's okay to use the local version generator for the loaded application state values
-    // As long as the endpoint_state has zero generation.
-    // It will get updated as a whole by handle_major_state_change
-    // via do_apply_state_locally when (remote_generation > local_generation)
```
yes, I was overzealous with my removing of stuff.
I'll resurrect it

bhalevy replied:
comment restored in v10

bhalevy commented on service/storage_service.cc:
```diff
@@ -669,7 +670,14 @@ future<> storage_service::topology_state_load() {
+            gms::loaded_endpoint_state st;
+            const auto& topo = tmptr->get_topology();
+            // node must exist in topology if it's in tmptr->get_all_endpoints
+            const auto* node = topo.find_node(e);
```
why?  dereferencing a null pointer, that should never happen, would result in a crash too.
anyhow I can use `get_node` that was introduced recently, and it would throw a runtime_error is the node was not found.

kbr-scylla replied:
IMO on_internal_error is better than regular runtime_error --- so we get an easier to debug failure in CI if it happens

bhalevy replied:
done in v10

bhalevy commented on service/storage_service.cc:
```diff
@@ -677,6 +677,11 @@ future<> storage_service::topology_state_load() {
             st.tokens = boost::copy_range<std::unordered_set<dht::token>>(tmptr->get_tokens(e));
             st.opt_dc_rack = node->dc_rack();
+            // Save tokens, not needed for raft topology management, but needed by legacy
+            // Also ip -> id mapping is needed for address map recreation on reboot
```
This would be ideal.
@gusev-p please confirm

kbr-scylla replied:
Nope, it's not true -- `sync_raft_topology_nodes` is only removing stuff from gossiper, never adding to it.
So please ignore that question

(but my remark is still true: id -> ip mapping doesn't rely on presence of tokens)

bhalevy replied:
revised comment in v10

bhalevy commented:
tested locally with `repair_based_node_operations_test.py::TestRepairBasedNodeOperations::{test_replace_node_after_cluster_restart,test_ignore_dead_nodes_for_replace_option_multi_dc,test_ignore_dead_nodes_for_replace_option_multi_rack}` dtests
with and without --experimental-features=consistent-topology-changes
command used:
```
export DTEST_REQUIRE=disabled CASSANDRA_DIR=$HOME/dev/scylla/build/dev SCYLLA_UNIFIED_PACKAGE=$HOME/dev/scylla/build/dev/dist/tar/scylla-unified-5.5.0~dev-0.20240226.ddeba190146d.x86_64.tar.gz && pytest --scylla-version=local_tarball -n 3 --experimental-features=consistent-topology-changes repair_based_node_operations_test.py::TestRepairBasedNodeOperations::{test_replace_node_after_cluster_restart,test_ignore_dead_nodes_for_replace_option_multi_dc,test_ignore_dead_nodes_for_replace_option_multi_rack}
```

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7020/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7020/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8199/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/160/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7020/)
#### Build Details:
- Duration: 2 hr 52 min
- Builder: i-027713b884acd4f19 (m5d.12xlarge)

bhalevy commented:
rebased

bhalevy commented:
@kbr-scylla , @asias please re-review.
I'd really like to merge this for 6.0

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7488/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7488/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/548/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8579/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7488/)
#### Build Details:
- Duration: 2 hr 5 min
- Builder: spider8.cloudius-systems.com

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -2121,18 +2121,20 @@ void gossiper::build_seeds_list() {
-        co_return;
+    }
+    if (ep == inet_address{}) {
+        on_internal_error(logger, format("Attempt to add {} with my null inet_address as saved endpoint", host_id));
```
s/my //

bhalevy replied:
fixed in v11

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -2683,20 +2694,19 @@ future<> storage_service::join_cluster(sharded<db::system_distributed_keyspace>&
         auto tmptr = co_await get_mutable_token_metadata_ptr();
         for (const auto& [host_id, st] : loaded_endpoints) {
-            if (st.endpoint == get_broadcast_address()) {
+            if (host_id == my_host_id()) {
```
Why did you reverse the cases? Previously the host_id conflict case would cause on_internal_error and the IP conflict case would cause a remove_endpoint, now it's reversed.

The previous logic made more sense to me.

I think the IP conflict case can actually happen, although it's very exotic. Suppose we have node A with IP X and node B with IP Y, first node B changes its IP to Z, then node A may change its IP to Y.

bhalevy replied:
makes sense.  I'll revert the order to the original one.
(and for the record, I hate the fact system.peers uses the ip address as key, rather than the host_id.  it's backward and we should get rid of that - maybe when we get rid of vnodes.  the sooner the better)

bhalevy replied:
unreversed in v11

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -2121,18 +2121,20 @@ void gossiper::build_seeds_list() {
 }
 
-future<> gossiper::add_saved_endpoint(locator::host_id host_id, inet_address ep, permit_id pid) {
+future<> gossiper::add_saved_endpoint(locator::host_id host_id, gms::loaded_endpoint_state st, permit_id pid) {
```
` gossiper: add_saved_endpoint: set dc and rack `

I don't see the logic which sets DC and Rack in this commit

bhalevy replied:
Odd.  It was probably got dropped by mistake when rebasing.
This PR had too many versions....
I'll add it back of course (and retest)

bhalevy replied:
restored in v11

kbr-scylla commented on api/storage_service.cc:
```diff
@@ -103,7 +103,7 @@ static bool any_of_keyspaces_use_tablets(const http_context& ctx) {
 
 locator::host_id validate_host_id(const sstring& param) {
```
` locator: host_id_or_endpoint: keep value as variant `

TBH I really don't like this commit, I think it's a bit controversial.

Is it actually needed to achieve the goal of this PR?

bhalevy replied:
why don't you like it?
why is it controversial?
That said, I might get around it and use the current class as it if it get in the way for pushing this PR through

bhalevy replied:
IIRC without this patch there is a problem resolving the ip address of ignored nodes for replace and we get the `Host ID {} not found in the cluster` exception since the hosts are not in token_metadata topology yet.

bhalevy replied:
kept in v11 as per the above

kbr-scylla commented:
Also @bhalevy please ping me if I don't review for more than 2 days

kbr-scylla commented:
ping @bhalevy 
I'm for pushing this through too

bhalevy commented on db/system_keyspace.cc:
```diff
@@ -1793,6 +1774,50 @@ future<std::unordered_map<gms::inet_address, locator::host_id>> system_keyspace:
+        }
+        auto host_id = locator::host_id(row.get_as<utils::UUID>("host_id"));
+        if (row.has("tokens")) {
+            st.tokens = decode_tokens(deserialize_set_column(*topology(), row, "tokens"));
```
oops, this was copy-pasted from `system_keyspace::load_topology_state`, but obviously it's wrong here.
Will fix.

bhalevy replied:
fixed in v11

bhalevy commented on db/system_keyspace.cc:
```diff
@@ -1793,6 +1774,50 @@ future<std::unordered_map<gms::inet_address, locator::host_id>> system_keyspace:
+            st.tokens = decode_tokens(deserialize_set_column(*topology(), row, "tokens"));
+            if (st.tokens.empty()) {
+                slogger.error("load_endpoint_state: node {}/{} has tokens column present but tokens are empty", host_id, ep);
+                continue;
```
An empty TOKENS is state is considered invalid, while a missing one might be not.
It's up to the caller to decide what do about it.
Currently it's not checked, but I think we should ignore it similar to `join_cluster` that adds nodes to `loaded_endpoint` only if they own tokens.

bhalevy replied:
fixed in v11 (ignoring loaded endpoints with empty tokens state)

bhalevy commented on db/system_keyspace.cc:
```diff
@@ -1793,6 +1774,50 @@ future<std::unordered_map<gms::inet_address, locator::host_id>> system_keyspace:
+            });
+            if (st.opt_dc_rack->dc.empty() || st.opt_dc_rack->rack.empty()) {
+                slogger.error("load_endpoint_state: node {}/{} has empty dc={} or rack={}", host_id, ep, st.opt_dc_rack->dc, st.opt_dc_rack->rack);
+                continue;
```
I think it's out of scope for this function.
It's up to the caller what to do with the raw metadata loaded here.

bhalevy replied:
cleaned up the setting of default dc/rack in v11

bhalevy commented on gms/gossiper.cc:
```diff
@@ -2134,12 +2134,16 @@ void gossiper::build_seeds_list() {
     }
```
In 6.0, with `peers_table_read_fixup`, peers should always have a valid HOST_ID.
If we backport this series we'll have take that host_id might be missing into account.

bhalevy commented:
In v11 ([f171478](https://github.com/scylladb/scylladb/commit/f1714784c2e6d48836371aa7a33bed67b672a14c)):
- rebased and retested
  - passed repair_based_node_operations_test.py::TestRepairBasedNodeOperations::{test_replace_node_after_cluster_restart,test_ignore_dead_nodes_for_replace_option_multi_dc,test_ignore_dead_nodes_for_replace_option_multi_rack} dtests with gossip, consistent-topology-changes, and tablets
- cleaned up series based on @kbr-scylla review comments.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7811/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7811/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8808/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7811/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/781/)
#### Build Details:
- Duration: 2 hr 56 min
- Builder: spider4.cloudius-systems.com

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -1441,7 +1441,7 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
         if (_sys_ks.local().bootstrap_complete()) {
```
`storage_service: prepare_replacement_info: load ignored nodes state`

One thing I don't understand about this commit.

> But since the down nodes have unknown status, they never go
through `handle_state_normal` and therefore they are not accounted
as normal token owners.

But in previous commit, you changed that:
`storage_service: topology_state_load: set local STATUS state using add_saved_endpoint`

Wasn't the point of that commit precisely to make sure that such down nodes appear as NORMAL, therefore, when a node starts the replace operation, it will learn about all NORMAL nodes through `handle_state_normal` --- including the ignored nodes? So we shouldn't have to give special treatment to ignored nodes.

We should insert **all** normal nodes to token ring when we boot, ignored or not.

Nodes that are normal but not ignored, should also get inserted to token ring. If not, we will calculate replicas incorrectly, like in https://github.com/scylladb/scylla-enterprise/issues/3518. It's not really about ignored nodes -- it's about us missing tokens of some nodes.

So I guess my question is, why do we need this `storage_service: prepare_replacement_info: load ignored nodes state` commit -- doesn't the STATUS commit already fix the issue?

bhalevy replied:
@kbr-scylla:

> `storage_service: prepare_replacement_info: load ignored nodes state`
> 
> One thing I don't understand about this commit.
> 
> > But since the down nodes have unknown status, they never go
> > through `handle_state_normal` and therefore they are not accounted
> > as normal token owners.
> 
> But in previous commit, you changed that: `storage_service: topology_state_load: set local STATUS state using add_saved_endpoint`
> 
> Wasn't the point of that commit precisely to make sure that such down nodes appear as NORMAL, therefore, when a node starts the replace operation, it will learn about all NORMAL nodes through `handle_state_normal` --- including the ignored nodes? So we shouldn't have to give special treatment to ignored nodes.

This happens only for consistent-topology-changes in `storage_service::topology_state_load`, not with gossip.
For the latter we need this patch to load the ignored nodes state in prepare_replacement_info.

> We should insert **all** normal nodes to token ring when we boot, ignored or not.

I agree.
But the replacing node doesn't know about them in advance since it doesn't have system.peers yet.
My original patch loaded all normal token owners in add_saved_endpoint, setting their status initially to shutdown, which is considered a NORMAL status, giving them a chance to become be marked alive when we successfully communicate with them.  But Asias objected to that and preferred the endpoint state to have unknown STATUS - and the consequence is that it stays hidden in gossip - hence the need to add the saved state after a shadow round for replacing node.

> Nodes that are normal but not ignored, should also get inserted to token ring. If not, we will calculate replicas incorrectly, like in [scylladb/scylla-enterprise#3518](https://github.com/scylladb/scylla-enterprise/issues/3518). It's not really about ignored nodes -- it's about us missing tokens of some nodes.
> 
> So I guess my question is, why do we need this `storage_service: prepare_replacement_info: load ignored nodes state` commit -- doesn't the STATUS commit already fix the issue?

See above.

I feel like we're going in circles for this PR (that now needs to be rebased yet again).
It can be simplified with the approach I mentioned above - that is to always add the saved state as NORMAL/shutdown, if @asias removes his objection to that.
I agree with you that it would be safer this way, since is the best we know about those endpoints, given their state in system.peers (and since we don't save their last STATUS persistently)

bhalevy commented:
In v12 ([655d624](https://github.com/scylladb/scylladb/commit/655d624e0198a8923e6744a9c91d4141da431b82)):
- rebased + retested
- split last patch and clarified commitlog message
  - plus, added comment in code to explain the handling of ignored nodes state in `join_token_ring` when raft topology changes are disabled.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7982/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7982/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/948/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8976/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7982/)
#### Build Details:
- Duration: 2 hr 20 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
Removed backport-5.4 tag since it is too risky
